Add UnitTests for scatterScalarResponse and scatterResidual with the HessianVec EvaluationType.

kliegeois opened the pull request:
This PR follows and is based on Albany [PR 634](https://github.com/SNLComputation/Albany/pull/634) and  [PR 637](https://github.com/SNLComputation/Albany/pull/637).

This PR adds UnitTests for `scatterScalarResponse` and `scatterResidual` with the `HessianVec EvaluationType`.

The added tests rely on:
- the construction of stand-alone scatter evaluators,
- setting the required dependent fields,
- evaluating the evaluators,
- the comparison of the output Thyra vectors with the expected vectors computing the norm of their difference.

mperego commented on tests/unit/evaluators/EvalTestSetup.hpp:
```diff
@@ -92,11 +92,10 @@ createTestLayoutAndBasis(Teuchos::RCP<Albany::Layouts> &dl,
 }
 
-void createTestMapsAndWorksetConns(
+Albany::WorksetConn createTestMapsAndWorksetConns(
```
Why did you choose to return `wsLocalElNodeEqID`. I don't understand why `wsLocalElNodeEqID` should be handled differently than `wsGlobalElNodeEqID`.

mperego reviewed (commented):
Looks good.  Can you please add comments to explain what are you testing?

kliegeois commented:
@mperego, I added scatter tests with `tensor rank = 1` and comments. I cleaned the EvalTestSetup.hpp file too.

The tests passed:

```kim@gimli:~/dev/Albany_all/AlbanyB$ ctest -R Albany
Test project /home/kim/dev/Albany_all/AlbanyB
      Start 157: Albany_Serial_Unit_Test
 1/10 Test #157: Albany_Serial_Unit_Test .................................   Passed    0.38 sec
      Start 158: Albany_Parallel_Unit_Test
 2/10 Test #158: Albany_Parallel_Unit_Test ...............................   Passed    0.37 sec
      Start 159: Albany_Serial_GatherSolution_Unit_Test
 3/10 Test #159: Albany_Serial_GatherSolution_Unit_Test ..................   Passed    0.36 sec
      Start 160: Albany_Parallel_GatherSolution_Unit_Test
 4/10 Test #160: Albany_Parallel_GatherSolution_Unit_Test ................   Passed    0.43 sec
      Start 161: Albany_Serial_GatherDistributedParameters_Unit_Test
 5/10 Test #161: Albany_Serial_GatherDistributedParameters_Unit_Test .....   Passed    0.36 sec
      Start 162: Albany_Parallel_GatherDistributedParameters_Unit_Test
 6/10 Test #162: Albany_Parallel_GatherDistributedParameters_Unit_Test ...   Passed    0.38 sec
      Start 163: Albany_Serial_ScatterScalarResponse_Unit_Test
 7/10 Test #163: Albany_Serial_ScatterScalarResponse_Unit_Test ...........   Passed    0.39 sec
      Start 164: Albany_Parallel_ScatterScalarResponse_Unit_Test
 8/10 Test #164: Albany_Parallel_ScatterScalarResponse_Unit_Test .........   Passed    0.53 sec
      Start 165: Albany_Serial_ScatterResidual_Unit_Test
 9/10 Test #165: Albany_Serial_ScatterResidual_Unit_Test .................   Passed    0.39 sec
      Start 166: Albany_Parallel_ScatterResidual_Unit_Test
10/10 Test #166: Albany_Parallel_ScatterResidual_Unit_Test ...............   Passed    0.48 sec

100% tests passed, 0 tests failed out of 10

Total Test time (real) =   4.08 sec
```

I will add gather tests with `tensor rank = 1` in another PR if it is fine.

mperego reviewed (approved):
Thanks @kliegeois. Looks great! I'm glad that the Hessian AD is now properly unit tested.

The only think I would add is a comment for each test explaining how you test the Hessian.. i.e. what values you are setting for the parameters, the solution,  the p and x directions, etc.,  and what are the expected hessian vec products. In this way it's easier for a developer to follow the code.

kliegeois commented:
> Thanks @kliegeois. Looks great! I'm glad that the Hessian AD is now properly unit tested.
> 
> 
> 
> The only think I would add is a comment for each test explaining how you test the Hessian.. i.e. what values you are setting for the parameters, the solution,  the p and x directions, etc.,  and what are the expected hessian vec products. In this way it's easier for a developer to follow the code. 

@mperego Thanks for your review, I will extend the comments as suggested.

kliegeois commented:
> Thanks @kliegeois. Looks great! I'm glad that the Hessian AD is now properly unit tested.
> 
> The only think I would add is a comment for each test explaining how you test the Hessian.. i.e. what values you are setting for the parameters, the solution, the p and x directions, etc., and what are the expected hessian vec products. In this way it's easier for a developer to follow the code.

@mperego do the comments included in the commit [
2611d4b](https://github.com/SNLComputation/Albany/pull/639/commits/2611d4ba9027090ad0fc50f46062c1bd4d31f5d9) fulfill the purpose listed in your comment?

mperego reviewed (approved):
Thanks a lot @kliegeois, I think this make the tests easy to understand.
